README.md
# myanmar-notation

A [Node.js](#nodejs) module and a [library](#browser) which converts the given number into words in Myanmar. Following written rule and add the creaky tone on number places `10^5`, `10^6` and `10^7`. Check [live Application][demo] is available...

[![workflows-badge]][workflows]
[![travis-badge]][travis]
[![npm-badge]][npm]
[![webpack-badge]][latest-min]
![test-mocha]

Normally "ပေါင်း" in Myanmar translated "plus", however the same word using in a large number of amount dedicate multiply. For example
သိန်းတစ်သောင်း is `1 0000 00000` and can also/usually pronounced သိန်းပေါင်း တစ်သောင်း which mean "`(100000*10000)`", therefor "သိန်းတစ်သောင်း" and "သိန်းပေါင်း တစ်သောင်း" has dedicated the same amount. This module is using "ပေါင်း" when the dedicated amount is larger than the subsequent scale name "ခု", "ဆယ်", "ရာ", "ထောင်", "သောင်း", "သိန်း", "သန်း" and "ကု​ဋေ​".

It can be a bit confusing between "သိန်းတစ်သိန်း" and "တစ်သိန်း" as of spoken usually does

> "ငွေ သိန်းတစ်သိန်းလောက်  ချေးပေးလို့ရမလား ခင်ဗျာ"

they probabbly just wanted "100000",
but there are entirely different as "သိန်းတစ်သိန်း" mean "10000 000000" and "တစ်သိန်း" mean "100000".

There are also difference positioning scale name, primarly it is followed by the given number

> 10: "တစ်ဆယ်" 20: "နှစ်ဆယ်" 10000000: "တစ်ကု​ဋေ​"

However using this primarly structure in large amount seem very weird, eg.100000000: "တစ်ထောင်သိန်း", "တစ်ရာသန်း", "တစ်ဆယ်ကု​ဋေ​", therefore in such amount the scale name lead the number to take away this weirdness.

> 100000000: "သိန်းတစ်ထောင်", "သန်းတစ်ရာ", "ကု​ဋေ​တစ်ဆယ်"

## Feature

- [x] Decimals are rounded (floor)
- [x] Query `get(2700)`, `get('၂၇၀၀')`, `get('27,000,000.00')` ,`get('5.23e+8')` is flexible
- [x] [Demo][demo]
- [x] mocha
- [x] webpack

## Usage

### Node.JS

`npm i myanmar-notation` to install, and can be `require` or `import` from your Node.JS application. However `myanmar-notation` is assuming that one day npm might force us to seperate ES and CommonJS module. Therefore, it is a good practice to start coding Node.JS application using ES6 module.

```js
// ES6
import notation from 'myanmar-notation';
import {get} from 'myanmar-notation';

notation.get(2700);
// notation.multiplication(5);
// notation.keep(123);
// notation.turn('၄၅၆');
get(12345678);

// CommonJS
const notation = require('myanmar-notation');
const {get} = require('myanmar-notation');

notation.get(2700);
get(12345678);
```

### Browser

Include the file `myanmar-notation@latest/min.js` in your web application. It is available on [UNPKG][unpkg].

```html
<script src="https://unpkg.com/myanmar-notation@latest/min.js"></script>
<script>
let notation = window.myanmarNotation;
</script>
```

## API

- [multiplication](#multiplication)
- [keep](#keep)
- [turn](#turn)
- [get](#get)

### Multiplication

the `notation.multiplication()` construct the given len to zero ending, it takes 2 arguments `(len, head='1')`, second argument is optional and return the result as string.

```js
notation.multiplication(5,'10');
> "100000"
```

### Keep

the `notation.keep()` convert the given string(number) to Burmese number and return the result as string.

```js
notation.keep(1234567);
> "၁၂၃၄၅၆၇"

notation.keep('a567');
> "a၅၆၇"
```

### Turn

the `notation.turn()` convert the given string(number) to English number and return the result as string.

```js
notation.turn('၁၂၃၄၅၆၇');
> "1234567"

notation.turn('123၅၆၇');
> "123567"
```

### get

the `notation.get()` return an object if there is a sense in get query, otherwise empty Object return.

```js
notation.get(2700);
// return
>  {
    "number": "၂၇၀၀",
    "digit": "2700",
    "notation": [
      {
        "sense": "နှစ်ထောင့်ခုနစ်ရာ"
      }
    ]
  }

notation.get(12345678);
// return
>  {
    "number": "၁၂၃၄၅၆၇၈",
    "digit": "12345678",
    "notation": [
      {
        "sense": "သိန်းတစ်ရာ့နှစ်ဆယ့်သုံး လေးသောင်းငါးထောင့်ခြောက်ရာ့ခုနစ်ဆယ့်ရှစ်",
        "rule": 5,
        "size": 8,
        "list": [
          "123",
          "45678"
        ]
      },
      {
        "sense": "သန်းတစ်ဆယ့်နှစ် သုံးသိန်းလေးသောင်းငါးထောင့်ခြောက်ရာ့ခုနစ်ဆယ့်ရှစ်",
        "rule": 6,
        "size": 8,
        "list": [
          "12",
          "345678"
        ]
      },
      {
        "sense": "တစ်ကု​ဋေ​ နှစ်သန်းသုံးသိန်းလေးသောင်းငါးထောင့်ခြောက်ရာ့ခုနစ်ဆယ့်ရှစ်",
        "rule": 7,
        "size": 8,
        "list": [
          "1",
          "2345678"
        ]
      }
    ]
  }
```

## License

![shield-license]

[demo]: https://khensolomon.github.io/myanmar-notation/
[workflows-badge]: https://github.com/khensolomon/myanmar-notation/workflows/Node/badge.svg
[workflows]: https://github.com/khensolomon/myanmar-notation/actions/workflows/node.yml
[test-mocha]: https://img.shields.io/badge/test-mocha-green.svg?longCache=true
[webpack-badge]: https://img.shields.io/badge/webpack-yes-green.svg?longCache=true
[latest-min]: https://unpkg.com/myanmar-notation@latest/min.js
[unpkg]: https://unpkg.com/
[travis-badge]: https://travis-ci.com/khensolomon/myanmar-notation.svg
[travis]: https://travis-ci.com/khensolomon/myanmar-notation
[npm-badge]: https://img.shields.io/npm/dt/myanmar-notation.svg
[npm]: https://www.npmjs.com/package/myanmar-notation
[shield-license]: https://img.shields.io/github/license/khensolomon/myanmar-notation?style=social


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## 1.4.6

- Improve
  - Docs
  - Usage
  - API
  - Feature
- Change
  - keep and turn API
- engines, Node.JS Compatibility
- Added
  - Code of Conduct
  - Security
- devDependencies
  - @babel/core: 7.13.13
  - @babel/preset-env: 7.13.12
  - css-loader: 5.2.0
  - mini-css-extract-plugin: 1.4.0
  - mocha: 8.3.2
  - sass: 1.32.8
  - sass-loader: 11.0.1
  - typescript: 4.2.3
  - vue-resource: 1.5.2

## 1.4.5

Fast and clean

- Milestone
  - [x] 1 ၁ တစ်
  - [x] 10 ၁၀ တစ်ဆယ်
  - [x] 100 ၁၀၀ ရာ
  - [x] 1000 ၁၀၀၀ ထောင်
  - [x] 10000 ၁၀၀၀၀ သောင်း
  - [x] 100000 ၁၀၀၀၀၀ သိန်
  - [x] 1000000 ၁၀၀၀၀၀၀ သန်း
  - [x] 10000000 ၁၀၀၀၀၀၀၀ ကုဋေ

- Structure

```js
[2/6] 1 00000 = (100000*1)
တစ်သိန်း
သိန်းပေါင်း တစ်

[2/7] 12 00034 = (100000*12+34)
တစ်ဆယ့်နှစ် သိန်း နှင့် သုံးဆယ့်လေး
သိန်းပေါင်း တစ်ဆယ့်နှစ် နှင့် သုံးဆယ့်လေး

[2/8] 123 00000 = (100000*123)
တစ်ရာ့နှစ်ဆယ့်သုံး သိန်း
သိန်းပေါင်း တစ်ရာ့နှစ်ဆယ့်သုံး

[2/9] 1234 00000 = (100000*1234)
တစ်ထောင့်နှစ်ရာ့သုံးဆယ့်လေး သိန်း
သိန်းပေါင်း တစ်ထောင့်နှစ်ရာ့သုံးဆယ့်လေး

[2/10] 12345 00000 = (100000*12345)
တစ်သောင်းနှစ်ထောင့်သုံးရာ့လေးဆယ့်ငါး သိန်း
သိန်းပေါင်း တစ်သောင်းနှစ်ထောင့်သုံးရာ့လေးဆယ့်ငါး

[3/11] 1 00000 00000 = (100000*100000)
uon-r
သိန်းပေါင်း တစ်သိန်း

[3/12] 12 00000 00000 = (100000*1200000)
uon-r
သိန်းပေါင်း တစ်ဆယ့်နှစ်သိန်း
သိန်းပေါင်း သိန်းတစ်ဆယ့်နှစ်

[3/15] 12 00000 00000 = (100000*1200000000)
uon-r
သိန်းပေါင်း သိန်းတစ်သောင်းနှစ်ထောင်

[4/16] 1 00000 00000 00000 = (100000*100000*100000)
uon-r
သိန်းပေါင်း သိန်း တစ်သိန်း

[4/20] 10000 00000 00000 00000 = (100000*100000*1000000000)
uon-r
သိန်းပေါင်း သိန်း သိန်းတစ်သောင်း

[5/21] 1 00000 00000 00000 00000 = (100000*100000*100000*100000)
uon-r
သိန်းပေါင်း သိန်း၊ သိန်း တစ်သိန်း

Nested-simple/Nested-complex
[5/21] 1 20000 30000 40000 50000 = (100000*100000*100000*120000+(100000*3000040005)).toLocaleString()
.toLocaleString('fullwide', {useGrouping:false});
uon-r
သိန်းပေါင်း သုံးသောင်း၊ လေးသောင်း တစ်သိန်း နှစ်သောင်း

သိန်းပေါင်း တစ်သိန်း နှစ်သောင်း

သိန်း နှစ်ထောင်

(100000*100000*12000000000+(100000*30000400000)+50000).toLocaleString('fullwide', {useGrouping:false})
(100000*100000*12000000000+(100000*3000040000)+50000).toLocaleString('fullwide', {useGrouping:false})
```

## 1.4.3

Good days

## 1.3.4

now in Typescript & build test

- Added
  - [x] Creaky tone

## 1.0.6

Straight JS final release

## 1.0.4

Improvement

## 1.0.3

Flexible get() query


notation.v1.0.4.js
module.exports = class Notation {
  constructor() {
    this.config={
      name:["သုည","တစ်","နှစ်","သုံး","လေး","ငါး","ခြောက်","ခုနစ်","ရှစ်","ကိုး"],
      digit:["ဝ","၁","၂","၃","၄","၅","၆","၇","၈","၉"],
      // 1:"ခု",
      tone:{1:"", 2:"ဆယ်", 3:"ရာ", 4:"ထောင်", 5:"သောင်း", 6:"သိန်း", 7:"သန်း", 8:"ကု​ဋေ​"},
      creaky:{2:"ဆယ့်",3:"ရာ့",4:"ထောင့်"},
      conjunction:{space:" ",comma:"၊ ",and:"နှင့်",plus:"ပေါင်း",over:"ကျော်",times:"ကြိမ်"},
      task:[6,7,8]
    }
  }
  configName() {
    return this.config.name;
  }
  configDigit() {
    return this.config.digit;
  }
  configTone() {
    return this.config.tone;
  }
  configCreaky() {
    return this.config.creaky;
  }
  configConjunction() {
    return this.config.conjunction;
  }
  get(query) {
    let q=Math.floor(Number(query.toString().replace(/,\s?/g, ""))).toString(),t=0;
    if (isNaN(q)){
      q=this.getUni(query),t=1;
      if (q<1 || isNaN(q)) return false;
    }
    return this.getResult(q,t);
  }
  getUni(query) {
    let n = query.toString().split('').map(k=> {
      return (this.config.digit.indexOf(k) >= 0)?this.config.digit.indexOf(k):k;
    }).filter(function(e){return e == 0 || e}).join('');
    return Math.floor(Number(n.replace(/,\s?/g, ""))).toString();
  }
  getResult(q,t) {
    let n = this.requestFormat(q);
    return {
      Numeric:(t)?n:this.requestDigit(n),
      Notation:this.config.task.map(k=>{
        return this.requestTask(q,k);
      }).filter(function(e){return e})
    };
  }
  requestTask(q,s) {
    if (q.length <= s) {
      if (this.config.task.length) {
        this.config.task=[];
        return {
          // sense:this.requestCreakyTone(q)
          sense:this.requestWrittenTone(q)
        }
      }
      return false;
    }
    let raw = this.strSeparate(q,s);
    let rawCount = raw.length;
    let rawEnd = raw.slice(-1)[0];
    let rawEndCount = rawEnd.length;

    let ruleExtract = rawEndCount;
    let rulePrime = rawCount;
    let ruleMax = Object.keys(this.config.tone).length;
    let rawSense=[];

    // let Measure=this.config.tone[s];
    // let ConjunctionPlus=this.config.conjunction.plus;
    // let PrefixeTimes='';
    // let SuffixeMeasure='';

    let examNotation=false;

    if (rawCount < 3) {
      if (s == rawEndCount) {
        ruleExtract=rawCount;
        // prime, base, baseName, rest, restName, tail, tailName
        rawSense.push(this.requestPrime(rulePrime,s)+this.requestCreakyTone(q,0,ruleExtract));
        rawSense.push(this.requestCreakyTone(q,rulePrime, s+1));
        rawSense.push(this.requestWrittenTone(q,s+1, q.length));
        // examNotation={1:21};
      } else {
        if (rawCount > rawEndCount) {
          ruleExtract=rawCount;
        } else {
          ruleExtract=rawEndCount+1;
        }
        rulePrime=1;
        rawSense.push(this.requestPrime(rulePrime,s,true)+this.requestCreakyTone(q,0,ruleExtract));
        rawSense.push(this.requestCreakyTone(q,ruleExtract, q.length));
        // examNotation={1:22};
      }
    } else {
      if (s == rawEndCount) {
        ruleExtract=rawCount;
        let rawOver='';
        if (Number(q.substring(ruleExtract, q.length))){
          rawOver=this.config.conjunction.over;
        }
        rawSense.push(this.requestPrime(rulePrime,s)+this.requestCreakyTone(q,0,ruleExtract)+rawOver);
        // examNotation={1:31};
      } else {
        ruleExtract=rawCount-1+rawEndCount;
        rulePrime=rawCount-1;
        // ruleExtract=rawCount;
        if (ruleExtract > ruleMax){
          ruleExtract=rawEndCount-3;
          rulePrime=rawCount;
        }
        let rawOver='';
        // if (Number(q.substring(ruleExtract, q.length))){
        //   rawOver=this.config.conjunction.over;
        // }
        let leftOver = q.substring(ruleExtract, q.length);
        if (Number(leftOver)){
          rawOver=this.config.conjunction.over;
          // leftOver = q.substring(ruleExtract, q.length-s);
          leftOver = q.substring(ruleExtract, q.length-s);
        }
        // 1234567000000==100000*12300000+4567000000
        // 1234567000000==100000*123 45670
        // 1234567000000000==100000*12345 600000+7000 000000
        rawSense.push(this.requestPrime(rulePrime,s)+this.requestCreakyTone(q,0,ruleExtract)+rawOver);
        examNotation={
          working:true,
          a1:q,
          raw:raw,
          b11:q.substring(0,ruleExtract),
          b12:q.substring(ruleExtract,s+2),
          b13:q.substring(ruleExtract,q.length),
          // sample:q.substring(ruleExtract,8),
          a3:leftOver,
          s:s,
          ruleExtract:ruleExtract,
          cut:s+ruleExtract,
          a4:leftOver.length,
          a5:this.strSeparate(q.substring(ruleExtract, q.length),ruleExtract)
        };
      }
    }

    return {
      sense:this.createString(rawSense),
      exam: examNotation
    };
  }
  requestObject(str,callback) {
    let strCount = str.length, e=[];
    // let testing = this.config.tone.reverse()
    // let strReverse = str.reverse(); index += to, ++index
    for (let index = 0; index < strCount;++index){
      let position = strCount - index;
      let digit = str[index], next = index+1;
      let tone=this.config.tone.hasOwnProperty(position)?this.config.tone[position]:'';

      if (digit > 0)e.push({
        name:this.config.name[digit],
        tone:callback(position,next,tone)
      });
    }
    return e;
  }
  requestCreakyTone(num,from,to) {
    let str = this.strIntersect(num,from,to);
    return this.requestObject(str,(position,next,tone)=>{
      if (str.hasOwnProperty(next) && str[next] > 0){
        return this.config.creaky.hasOwnProperty(position)?this.config.creaky[position]:tone;
      }
      return tone;
    }).map((k,index)=>{
      return k.name+k.tone;
    }).join(' ')
  }
  requestWrittenTone(num,from,to) {
    let str = this.strIntersect(num,from,to);
    return this.requestObject(str,(position,next,tone)=>{
      return tone;
    }).map((k,index,e)=>{
      if ((e.length - 2) == index) {
        return k.name+k.tone+this.config.conjunction.and;
      }
      return k.name+k.tone;
    }).join(' ');
  }
  requestCreakyTail(num,from,to) {
    let str = this.strIntersect(num,from,to);
    return this.requestObject(str,(position,next,tone)=>{
      if (str.hasOwnProperty(next) && this.config.creaky.hasOwnProperty(position) && str[next] > 0){
        return {normal:tone, creaky:this.config.creaky[position]};
      }
      return {normal:tone}
    }).map((k,index,e)=>{
      if ((e.length - 2) == index) {
        return k.name+k.tone.hasOwnProperty('creaky')?k.name+k.tone.creaky:k.name+k.tone.normal;
      }
      return k.name+k.tone.normal;
    }).join(' ');
  }
  requestDigit(n) {
    return n.split('').map(k=>{
      return this.config.digit.hasOwnProperty(k)?this.config.digit[k]:k;
    }).join('');
  }
  requestFormat(n) {
    return n.replace(/(\d)(?=(\d{3})+(?!\d))/g, "$1,");
  }
  requestPrime(l,k,s) {
    let i=0, e=[];
    while( i < l) { e.push(this.config.tone[k]); i++;}
    if (s)e.push('');
    return e.join(this.config.conjunction.plus+this.config.conjunction.space);
  }
  createString(n) {
    return n.filter(function(e){return e}).join(this.config.conjunction.and+this.config.conjunction.space)
  }
  strSeparate(str, n) {
    let e = new Array;
    for (let index=0; index < str.length; index += n) e.push(str.substr(index,n));
    return e;
  }
  strIntersect(str,from,to) {
    if (to){
      return str.substring(from, to).split('');
    } else{
      return str.split('');
    }
  }
}